### PR TITLE
Improvement/try-catch consensus

### PR DESCRIPTION
**Pull request by denis-trapeznikov**
One possible implementation of a nothrow consensus. Consensus interface is left intact (still working with blocks rather than pointers) but all the exception explicitly thrown are contained inside. In other words, program state monad is converted into that of zeroable value when control leaves the inwards of `Consensus` class.

Also discovered a very funny LGDR-699, it is also (probably) fixed here, since the whole fix took a replacement of one word in the code.

### CODE REVIEW

File: libs/ledger/include/ledger/consensus/consensus.hpp [hunk 1]
```diff
@@ -82,6 +82,7 @@ class Consensus final : public ConsensusInterface
   Consensus &operator=(Consensus const &) = delete;
   Consensus &operator=(Consensus &&) = delete;
 
+  // TODO (nobody): the only reason this function is public is it's used in a single POS test.
```
Comment (ejfitzgerald) [inline libs/ledger/include/ledger/consensus/consensus.hpp]: Can you create a JIRA issue to track this and update the code

File: libs/ledger/src/consensus/consensus.cpp [hunk 2]
```diff
@@ -261,13 +261,20 @@ bool Consensus::VerifyNotarisation(Block const &block) const
...
+        auto ordered_notarisation_keys = aeon_block.block_entropy.aeon_notarisation_keys;
 
-      return notarisation_->Verify(previous_notarised_block->hash,
-                                   block.block_entropy.block_notarisation,
-                                   ordered_notarisation_keys, threshold);
+        return notarisation_->Verify(previous_notarised_block->hash,
+                                     block.block_entropy.block_notarisation,
+                                     ordered_notarisation_keys, threshold);
+      }
+      catch (...)
```
Comment (ejfitzgerald) [inline libs/ledger/src/consensus/consensus.cpp:288]: Worth logging any error from here?

File: libs/ledger/src/consensus/consensus.cpp [hunk 3]
```diff
@@ -280,24 +287,25 @@ bool Consensus::VerifyNotarisation(Block const &block) const
...
-                                                          current.block_entropy.EntropyAsU64());
-
-  if (std::find(qualified_cabinet_weighted.begin(), qualified_cabinet_weighted.end(), identity) ==
-      qualified_cabinet_weighted.end())
+  Block            beginning_of_aeon;
+  try
+  {
+    beginning_of_aeon = GetBeginningOfAeon(current, chain_);
+  }
+  catch (...)
```
Comment (ejfitzgerald) [inline libs/ledger/src/consensus/consensus.cpp]: Again is it worth at least logging the error here?

File: libs/ledger/src/consensus/consensus.cpp [hunk 4]
```diff
@@ -610,6 +627,17 @@ NextBlockPtr Consensus::GenerateNextBlock()
     notarisation_->NotariseBlock(*ret);
   }
 
+#ifndef NDEBUG
+  try
+  {
+    GetBeginningOfAeon(*ret, chain_);
```
Comment (ejfitzgerald) [inline libs/ledger/src/consensus/consensus.cpp:650]: Is this supposed to be kept here?

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]
Looks like a nice change. Just a few comments. I think it is probably worth reviewing with @n-hutton which of the new catch blocks we want to log the exception message from

File: libs/ledger/src/consensus/consensus.cpp (hunk 4 above)
Comment (denis-trapeznikov) [reply libs/ledger/src/consensus/consensus.cpp:650]: See, earlier in the code, the weight is already calculated via `GetBlockGenerationWeight()`, so the beginning of aeon should not fail at this moment. Consequently, I've inserted this block as a debug-time assertion for that "should not".

File: libs/ledger/include/ledger/consensus/consensus.hpp (hunk 1 above)
Comment (denis-trapeznikov) [reply libs/ledger/include/ledger/consensus/consensus.hpp]: LGDR-698

File: libs/ledger/src/consensus/consensus.cpp [hunk 5]
```diff
@@ -66,6 +66,7 @@ namespace {
 
 constexpr char const *LOGGING_NAME = "Consensus";
 const std::size_t     DIGEST_LENGTH_BYTES{32};
+constexpr char const *unknown_exception = "something wenth wrong";
```
Comment (ejfitzgerald) [inline libs/ledger/src/consensus/consensus.cpp]:
```suggestion
constexpr char const *FATAL_ERROR = "Fatal error generated";
```

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]
Just a minor small suggestion looks okay. I can see that there is a little bit of feature creep on this PR

**Review by ejfitzgerald — APPROVED** [review]

**Pull request merged**